firefoxThemes/resources.js

```diff
@@ -1,45 +1,209 @@
-/*Location to retrieve resources for each waifu theme*/
+/*Globals*/
+//Enum for the different Mixed option states
+const mixedStates = {
+    NONE:0,
+    INITIAL:1,
+    TAB_CREATED:2
+};
+let mixedTabs;//Keep a record of each theme used on each tab
+/*---CLASSES---*/
+/*Class Goal: Holds theme data about all waifus*/
 class WaifuThemes {
-  constructor() {
-    this.initThemes();
-  }
-
-  /*Initialize Waifu Themes*/
-  initThemes() {
-    this.themes =
-      Object.entries(dokiThemeDefinitions).reduce((accum, [themeId, dokiTheme]) => ({
-        ...accum,
-        [themeId]: new Theme(
-          dokiTheme.information.displayName,
-          dokiTheme.information.imagePath,
-          dokiTheme.information.jsonPath,
-          dokiTheme,
-        )
-      }), {})
-
-  }
+    constructor() {
+        this.initThemes();
+    }
+    /*Initialize Waifu Themes*/
+    initThemes() {
+        this.themes =
+            Object.entries(dokiThemeDefinitions).reduce((accum, [themeId, dokiTheme]) => ({
+                ...accum,
+                [themeId]: new Theme(
+                    dokiTheme.information.displayName,
+                    dokiTheme.information.imagePath,
+                    dokiTheme.information.jsonPath,
+                    dokiTheme,
+                )
+            }), {})
+    }
+}
+/*Location to retrieve resources for each waifu theme*/
+class Theme {
+    constructor(name, image, json, themeDefinition) {
+        this.name = name;//Name of theme
+        this.image = image;//Relative link to waifu image file
+        this.json = json;//Relative link to browser theme file
+        this.definition = themeDefinition;
+        this.page = "/waifus/index.html";
+    }
+}
+/*---FUNCTIONS---*/
+/*Initialize Local Storage & custom new tab page*/
+async function startStorage() {
+    browser.storage.local.get("currentThemeId")
+        .then((storage) => {
+            const initStorage = {
+                waifuThemes: new WaifuThemes()
+            };
+            //Retrieve all themes if none exists in local storage
+            browser.storage.local.set(initStorage);
+            //Load browser theme
+            if (storage.currentThemeId) {
+                loadTheme(initStorage.waifuThemes.themes, storage.currentThemeId);
+            }
+            //When the browser first opens, redirect to custom new tab page
+            browser.tabs.update({loadReplace: true, url: "waifus/index.html"});
+        });
+}
+/*Update all new tabs with new waifu theme*/
+function updateTabs(msg) {
+    switch (msg.mixState){
+        case mixedStates.INITIAL:
+            SetupMixedUpdate();
+            break;
+        case mixedStates.TAB_CREATED:
+            MixedUpdate(msg);
+            break;
+        default:
+            MixTabCleanup();
+            NormalUpdate(msg);
+            break;
+    }
+}
+/*Initialize the Mixed feature*/
+function SetupMixedUpdate(){
```

Comment:
Just curious, are you following some example or documentation to add this functionality?

If so, could you please share?

Reply:
No. All of this comes straight from me noggin. I could explain my process for this feature if you wish.

### Inspiration
After the rework, I began testing the extension to see if everything was A-okay. What caught my interest was the random option, which randomly selects a theme for you. From here I thought, wouldn't it be cool to add to loads a random theme every time you create a new tab? Also I thought maybe someone might get tired of seeing the same waifu/theme every new tab.

### The Work
So to implement this new feature, I need to 
- listen for the current tab,
-  record the current theme for each tab, 
- listen to changes in the tab
- etc. 

With this planning cycle done, I navigate to MDN's [JavaScript API](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API) section to see what I can do with the tabs. If the tabs API does not provide the tools I need to accomplish the task, there is always a workaround that can be made. An example would be the previous version of this extension. I developed my own method for loading these themes... Only to find out later there was a [simpler method](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/chrome_url_overrides).

Using the information I obtained from the docs, I formulate a plan to implement it. Sleeping helps me think. 

And that's pretty much it.

Reply:
I'm guessing you think that way because of the excessive comments. I prefer to document what I can so when I come back I know what everything does.

Reply:
> I'm guessing you think that way because of the excessive comments

Probably, I was just curious about where you are finding out where these browser functionalities are hiding like: being able to replace the new tab action with the extensions new tab (and all of the boiler plate that came with that)

You've been adding a bunch of code, so I guess I'm trying to see if there is a place for boilerplate that you found. (Not saying boilerplate is bad, it's just stuff I don't want to write) 

For instance I just used [this guide](https://developer.chrome.com/docs/extensions/mv2/getstarted/) and [this boilerplate](https://github.com/chibat/chrome-extension-typescript-starter) to bootstrap the chrome extension.

Reply:
Oh yeah, I appreciate you documenting your thought process. However, you don't need to write me a book 😄 You could distill it to something like this:

> I looked through the [browser JavaScript api documentation](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API), and didn't find anything, so I made my own 
> ![I'll do it myself](https://media.tenor.com/images/4c397cf4cce3bb9a73b6e0665b55e8f5/tenor.gif)
> However, I later found out the [manifest spec](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/chrome_url_overrides) supports what I wanted to do better.